docs/src/architecture/09_architecture_decisions/0006-cursor-pegination-cat-data-service.md

```diff
@@ -0,0 +1,118 @@
+---
+    title: 0006 Cursor-based pagination for Catalyst Data Service
+    adr:
+        author: Oleksandr Prokhorenko
+        created: 05-Feb-2024
+        status:  proposed
+        extends:
+    tags:
+        - api
+---
+
+## Context
+
+Our API currently lacks a pagination mechanism, leading to inefficiencies in data retrieval processes,
+especially when dealing with large datasets.
+This absence has resulted in longer load times and
+a strained user experience, prompting the need for a scalable solution to manage data access and navigation effectively.
+
+## Assumptions
+
+* The assumption that cursor-based pagination will be universally supported and understood by all clients consuming our API.
+* Assuming our database and backend infrastructure can efficiently support cursor-based operations without
+ significant refactoring or performance degradation.
+* The expectation that data growth will continue at its current pace or accelerate,
+necessitating a robust solution.
+
+## Decision
+
+We are proposing to introduce a cursor-based pagination system to our API.
+This system will rely on ans enabling clients to efficiently fetch data in chunks and
+understand when more data is available:
+
+* `limit` - This establishes a restriction on the quantity of objects to be returned,
+with a minimum of 1 and a maximum of 100.
+* `starting_after`- the cursor to start returning results from,
+* `ending_before`- the cursor to stop returning results at.
+
+The JSON response structure will include fields such as
+
+* `data` - containing the requested data,
+* `has_more` - indicating whether more data is available,
+* `next_cursor`- the cursor to use for the next page of results,
```

Comment:
Do we need `next_cursor` and `prev_cursor`, given the data will have its `id`.  And we can use that for both the `starting_after` and `ending_before` fields in the subsequent query?

Reply:
Good question. Here are some thoughts:

Incorporating both `next_cursor/prev_cursor` and `starting_after/ending_before` fields can offer a flexible and robust pagination strategy that caters to various client needs and use cases:

**Flexibility for Clients:** By providing both options, we allow clients to choose the pagination method that best fits their requirements. For instance, some clients prefer cursor-based pagination for its simplicity and abstraction, especially when it is essential to navigating through a dataset without understanding its underlying structure. Others might opt for ID-based pagination (starting_after/ending_before) for more granular control, such as when implementing complex querying or navigating datasets with a straightforward, stable ordering.

**Backup Mechanism:** Offering both methods can serve as a backup mechanism. Suppose one pagination method encounters issues (e.g., performance concerns with large offsets in ID-based pagination or complexity in generating cursors). Clients can switch to the alternative method as a workaround in that case.

However, supporting both  methods also comes with **challenges**:

**Increased Complexity:** Implementing and maintaining both methods can increase the complexity of our backend. We'll need to ensure both mechanisms are correctly synchronized and consistently reflect the dataset's state, which might increase the development and testing overhead.
**Documentation and Learning Curve:** Providing clear and comprehensive documentation becomes crucial, as users need to understand the differences between the two methods, their use cases, and how to implement them correctly. This might steepen the learning curve for new users of our API.
**Potential for Inconsistencies:** Care must ensure that both pagination methods provide consistent data views, especially in highly dynamic datasets where the data changes frequently.

In conclusion, offering both `next_cursor/prev_cursor` and `starting_after/ending_before` can significantly increase the flexibility and robustness of our pagination strategy, accommodating a more comprehensive range of use cases and client preferences. However, weighing these benefits against the increased complexity and maintenance overhead is essential. We must consider our target audience, the nature of your dataset, and our team's capacity to support both methods effectively before deciding.

To summarize, I'd like to propose starting with `starting_after/ending_before` and postponing a decision on implementation on  `next_cursor/prev_cursor.`

What do you think, @stevenj ?

Reply:
I think all i was unclear on was how they were different,  if they are different values, then I have no problem with it.